presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,537 @@
+
+        if (filter != null) {
+            outputPositions = ensureCapacity(outputPositions, positionCount);
+        }
+        else {
+            outputPositions = positions;
+            outputPositionsReadOnly = true;
+        }
+
+        // account memory used by values, nulls, outputPositions, offsetVector, lengthVector, isNullVector
+        systemMemoryContext.setBytes(getRetainedSizeInBytes());
+
+        if (readOffset < offset) {
+            skip(offset - readOffset);
+        }
+
+        prepareForNextRead(positionCount, positions);
+
+        ensureValuesCapacity(positionCount, presentStream != null && nullsAllowed);
+
+        int streamPosition;
+
+        if (dataStream == null && lengthVector == null && isNullVector != null) {
+            streamPosition = readAllNulls(positions, positionCount);
+        }
+        else if (filter == null) {
+            streamPosition = readNoFilter(positions, positionCount);
+        }
+        else {
+            streamPosition = readWithFilter(positions, positionCount);
+        }
+
+        readOffset = offset + streamPosition;
+        return outputPositionCount;
+    }
+
+    private int readNoFilter(int[] positions, int positionCount)
+            throws IOException
+    {
+        int streamPosition = 0;
+        allNulls = false;
+
+        int offsetVectorIndex = 0;
+        offsetVector[offsetVectorIndex] = 0;
+        boolean isCharType = streamDescriptor.getOrcType().getOrcTypeKind() == OrcType.OrcTypeKind.CHAR;
+        int maxCodePointCount = getMaxCodePointCount(streamDescriptor.getOrcType());
+
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            if (position > streamPosition) {
+                skipData(streamPosition, position - streamPosition);
+                streamPosition = position;
+            }
+
+            if (isNullVector != null && isNullVector[position]) {
+                offsetVector[offsetVectorIndex + 1] = offsetVector[offsetVectorIndex];
+                offsetVectorIndex++;
+                nulls[i] = true;
+            }
+            else {
+                int offset = offsetVector[offsetVectorIndex];
+                int length = lengthVector[lengthIndex];
+                int truncatedLength = 0;
+                if (length > 0) {
+                    dataStream.next(data, offset, offset + length);
+                    truncatedLength = computeTruncatedLength(slice, offset, length, maxCodePointCount, isCharType);
+                }
+                offsetVector[offsetVectorIndex + 1] = offset + truncatedLength;
+                offsetVectorIndex++;
+                lengthIndex++;
+                if (nullsAllowed && isNullVector != null) {
+                    nulls[i] = false;
+                }
+            }
+            streamPosition++;
+        }
+        outputPositionCount = positionCount;
+        outputPositions = positions;
```

Comment:
The following comment is marked resolved but the code hasn't changed:

> Check out outputPositionsReadOnly logic in other readers and add it here.

Reply:
from other reader its used in read method and compact values. will follow up with you tomorrow on this

Reply:
@bhhari outputPositionsReadOnly indicates whether outputPositions array is owned by this reader or is a reference to positions array provided by the caller of read. Hence, `outputPositions = positions;` must be accompanied with `outputPositionsReadOnly = true;`. Here, you can either add `outputPositionsReadOnly = true;` or remove `outputPositions = positions;` entirely. This is already done in read():

```
       if (filter != null) {
            outputPositions = ensureCapacity(outputPositions, positionCount);
        }
        else {
            outputPositions = positions;
            outputPositionsReadOnly = true;
        }
```